model/work.py

```diff
@@ -0,0 +1,1603 @@
+            else:
+                pool.superceded = True
+
+    @classmethod
+    def restrict_to_custom_lists_from_data_source(
+            cls, _db, base_query, data_source, on_list_as_of=None):
+        """Annotate a query that joins Work against Edition to match only
+        Works that are on a custom list from the given data source."""
+
+        condition = CustomList.data_source==data_source
+        return cls._restrict_to_customlist_subquery_condition(
+            _db, base_query, condition, on_list_as_of)
+
+    @classmethod
+    def restrict_to_custom_lists(
+            cls, _db, base_query, custom_lists, on_list_as_of=None):
+        """Annotate a query that joins Work against Edition to match only
+        Works that are on one of the given custom lists."""
+        condition = CustomList.id.in_([x.id for x in custom_lists])
+        return cls._restrict_to_customlist_subquery_condition(
+            _db, base_query, condition, on_list_as_of)
+
+    @classmethod
+    def _restrict_to_customlist_subquery_condition(
+            cls, _db, base_query, condition, on_list_as_of=None):
+        """Annotate a query that joins Work against Edition to match only
+        Works that are on a custom list from the given data source."""
+        # Find works that are on a list that meets the given condition.
+        qu = base_query.join(LicensePool.custom_list_entries).join(
+            CustomListEntry.customlist)
+        if on_list_as_of:
+            qu = qu.filter(
+                CustomListEntry.most_recent_appearance >= on_list_as_of)
+        qu = qu.filter(condition)
+        return qu
+
+    def classifications_with_genre(self):
+        from classification import (
+            Classification,
+            Subject,
+        )
+        _db = Session.object_session(self)
+        identifier = self.presentation_edition.primary_identifier
+        return _db.query(Classification) \
+            .join(Subject) \
+            .filter(Classification.identifier_id == identifier.id) \
+            .filter(Subject.genre_id != None) \
+            .order_by(Classification.weight.desc())
+
+    def top_genre(self):
+        from classification import Genre
+        _db = Session.object_session(self)
+        genre = _db.query(Genre) \
+            .join(WorkGenre) \
+            .filter(WorkGenre.work_id == self.id) \
+            .order_by(WorkGenre.affinity.desc()) \
+            .first()
+        return genre.name if genre else None
+
+
+# Used for quality filter queries.
+Index("ix_works_audience_target_age_quality_random", Work.audience, Work.target_age, Work.quality, Work.random)
+Index("ix_works_audience_fiction_quality_random", Work.audience, Work.fiction, Work.quality, Work.random)
+
+
+class BaseMaterializedWork(object):
+    """A mixin class for materialized views that incorporate Work and Edition."""
+    pass
+
+
+class MaterializedWorkWithGenre(Base, BaseMaterializedWork):
+    p = dict(primary_key=True)
```

Comment:
We no longer need `index=True`?

Reply:
I believe `index=True` is only used by SQLAlchemy when initially creating the table. Since SQLAlchemy doesn't create the materialized view, it's redundant.

We do need the `primary_key=True` stuff because that's how SQLAlchemy knows how to look up an item in this table.